arkouda/alignment.py

```diff
@@ -383,61 +383,139 @@ def search_intervals(vals, intervals):
+        if not is_sorted(low):
+            raise ValueError("Intervals must be sorted in ascending order")
+        # verify intervals are non-overlapping
+        if not (low[1:] > high[:-1]).all():
+            raise ValueError("Intervals must be non-overlapping")
+
+        # Index of interval containing each unique value (initialized to -1: not found)
+        containinginterval = -ones(vals.size, dtype=akint64)
+        concat = concatenate((low, vals, high))
+        perm = argsort(concat)
+        # iperm is the indices of the original values in the sorted array
+        iperm = argsort(perm)  # aku.invert_permutation(perm)
+        boundary = vals.size + low.size
+        # indices of the lower bounds in the sorted array
+        starts = iperm[: low.size]
+        # indices of the upper bounds in the sorted array
+        ends = iperm[boundary:]
+        # which lower/upper bound pairs have any indices between them?
+        valid = ends > starts + 1
+        if valid.sum() > 0:
+            # pranges is all the indices in sorted array that fall between a lower and an uppper bound
+            segs, pranges = gen_ranges(starts[valid] + 1, ends[valid])
+            # matches are the indices of those items in the original array
+            matches = perm[pranges]
+            # integer indices of each interval containing a hit
+            hitidx = arange(valid.size)[valid]
+            # broadcast interval index out to matches
+            matchintervalidx = broadcast(segs, hitidx, matches.size)
+            # make sure not to include any of the bounds themselves
+            validmatch = (matches >= low.size) & (matches < boundary)
+            # indices of unique values found (translated from concat keys)
+            uvalidx = matches[validmatch] - low.size
+            # set index of containing interval for uvals that were found
+            containinginterval[uvalidx] = matchintervalidx[validmatch]
+        return containinginterval
+    elif isinstance(vals, (list, tuple)):
+        # argument validation for multi-array
+        if not isinstance(low, (list, tuple)) or not isinstance(high, (list, tuple)):
+            raise TypeError("intervals must be same type as vals")
+        if len(vals) != len(low) or len(vals) != len(high):
+            raise TypeError("Multi-array arguments must have same number of arrays")
+        if (
+            any([not isinstance(v, pdarray) for v in vals])
+            or any([not isinstance(lo, pdarray) for lo in low])
+            or any([not isinstance(hi, pdarray) for hi in high])
+        ):
+            raise TypeError("All elements of Multi-array arguments must be pdarrays")
+
+        valsize = set(v.size for v in vals)
+        lowsize = set(lo.size for lo in low)
+        highsize = set(hi.size for hi in high)
+        if len(valsize) != 1 or len(lowsize) != 1 or len(highsize) != 1:
+            raise TypeError("Multi-array arguments must be non-empty and have equal-length arrays")
```

Comment:
Do we really need everything to be the same size? Obviously, at index `i`, `vals[i].size == low[i].size == high[i].size` is required. But I am not sure there is a requirement for the various indexes to be the same size. Looking at the rest of the code, we are definitely relying on this, so I just wanted to check that we don't think there will be variance in sizes between elements.

Reply:
good point! this is definitely an assumption i was making that I'm not sure is necessarily the case. I was thinking about a comment @reuster986 made about [`MultiArray`s having all arrays be the same size ](https://github.com/Bears-R-Us/arkouda/pull/1532#discussion_r906307677)

But I'm not sure, what's right here. @reuster986, do u know if it's reasonable to expect all arrays stored in an a multi-array to have the same size?

Reply:
Based on @reuster986 input, I think this is good to go, resolving the thread.